### PR TITLE
Ds 3552 read only context and hibernate improvements

### PR DESCRIPTION
**Pull request by tomdesair**
This is a fix for https://jira.duraspace.org/browse/DS-3552

In short this PR:
* Fixes some recursion issues in GroupServiceImpl.isMember() and AuthorizeServiceImpl.isAdmin()
* Adds a "READ-ONLY" mode to the Context object which tells Hibernate to skip some checks when only doing read operations
* Adds additional caching for authorized actions and group membership when the Context is in READ-ONLY mode
* Fixes and improves the Hibernate 2nd Level Ehcache configuration

### CODE REVIEW

**Comment by tdonohue** [comment]
@tomdesair : according to Travis, this PR is currently failing a few Unit Tests under `org.dspace.core.ContextTest.testIsReadOnly()`

**Comment by tomdesair** [comment]
I'll check that tomorrow. But I'm following the idea of "create a PR when you already have something" ;-)

**Comment by tdonohue** [comment]
@tomdesair : 👍  Definitely do feel free to create PRs when you have something, as that is the recommended approach. I was just wanting to make sure you were aware of the Travis failures (as sometimes folks overlook those). So, no rush!

**Comment by tdonohue** [comment]
I realize this is still a work-in-progress, but wanted to note that there may be overlap between this effort and #1702, as both seem to be enhancing the performance of the `Context` object in relation to Hibernate.  I'd recommend also looking at that PR (specifically this commit https://github.com/DSpace/DSpace/pull/1702/commits/dc67a9a5b12871bb4f131c47e7830c69d0971152) to see if these are related.

**Comment by tomdesair** [comment]
This PR can be considered final, see the comments in https://jira.duraspace.org/browse/DS-3552

@tdonohue Note that a "clearCache()" was explicitly removed (again after being re-introducted to cope with harvesting large OAI sets) in this PR: https://github.com/DSpace/DSpace/pull/1495/files#diff-cbd2f003ca152958b516d0e949719ef3

**Comment by tdonohue** [comment]
Thanks @tomdesair. I missed the update over in the ticket.

Just to clarify, do you feel that this PR will have any affect on fixing https://jira.duraspace.org/browse/DS-3559? (This is the new ticket that #1702 was created to resolve). I'm just trying to understand whether DS-3552 and DS-3559 are related or not.

**Comment by tomdesair** [comment]
Since building the OAI index is a read-only operation on the database, it is possible that this PR can make it faster.

I'll try to do some testing tomorrow.

**Comment by pnbecker** [comment]
@tomdesair Thanks a lot for the work you put into this. I was looking to this issue as well and just found your PR. I have just read parts of your code changes and I really like the approach you took to solve the problems listed above. I hope I'll find time for a proper review and test.

**Comment by christian-scheible** [comment]
@tomdesair I tested your pull request for the OAI-Indexing. Your changes solves the problem with memory inefficiency and speed.
33000 items took about 10 minutes to index and there have been 336 items (which is pretty low) in the hibernate cache at the end of the indexing.

**Comment by tomdesair** [comment]
My last commit (7e1a0a1) fixes a memory-leak when running index-discovery. I've added it to this PR because the code for index-discovery already used the `uncacheEntity()` method.

**Comment by pnbecker** [comment]
For local reasons it was not so simple to test this. I just finished my tests and I'm satisfied.

I have almost two hours left for a review today. In case I do not finish my review today and someone else chips in, feel free to merge.

@tomdesair Thanks for your work on this one!

File: dspace/config/hibernate-ehcache-config.xml
Comment (pnbecker) [inline dspace/config/hibernate-ehcache-config.xml]: this must be added to the DSpace manual as well!

File: dspace/config/hibernate-ehcache-config.xml
Comment (pnbecker) [inline dspace/config/hibernate-ehcache-config.xml]: With 4.000 Collections we probably have more than 3.000 groups. ;-)

File: dspace-api/src/test/java/org/dspace/core/ContextTest.java
Comment (pnbecker) [inline dspace-api/src/test/java/org/dspace/core/ContextTest.java]: Do we? In the same commit you removed the check for being in read only mode when doing a database rollback.

File: dspace-api/src/main/java/org/dspace/core/HibernateDBConnection.java
Comment (pnbecker) [inline dspace-api/src/main/java/org/dspace/core/HibernateDBConnection.java]: We are preparing a cache that allows to delete EPersons even if they submitted items. In this case a submitter can be null, so please avoid a NPE here.

File: dspace-api/src/main/java/org/dspace/core/HibernateDBConnection.java
Comment (pnbecker) [inline dspace-api/src/main/java/org/dspace/core/HibernateDBConnection.java]: WorkSpace and WorkflowItems do not have any handles. May this lead to a problem?

**Review by pnbecker — COMMENTED** [review]
Overall this looks like a great improvement. Thanks for your hard work!

I have one main concern: Do we need to reset the cache whenever we go to read only mode, coming from one of the other two modes? I'm afraid of strange side effects when switching multiple times between read only and the other modes.

**Comment by hardyoyo** [comment]
@pnbecker it's not clear from the threading above whether you're satisfied with the state of this work, or if you need further changes before approving this PR. Since you did not actually approve the PR, I have to assume that you're not ready for this to be merged. I'm OK with merging this, so I'm going to approve it, but will let you make the final call on pressing the merge button.

**Review by hardyoyo — APPROVED** [review]

**Comment by pnbecker** [comment]
@hardyoyo I have one main concern: Do we need to reset the cache whenever we go to read only mode, coming from one of the other two modes? I'm afraid of strange side effects when switching multiple times between read only and the other modes. Beside that I only had small change request that should be solvable within 15 minutes all together.

File: dspace-api/src/main/java/org/dspace/core/HibernateDBConnection.java
Comment (tomdesair) [inline dspace-api/src/main/java/org/dspace/core/HibernateDBConnection.java]: No it won't lead to problems. Hibernate will only touch the handles if they have been retrieved from the database before. The `Utils.emptyIfNull()` will make sure we don't get any null pointer exceptions in case the handles collection is actually null.

File: dspace-api/src/main/java/org/dspace/core/Context.java
Comment (tomdesair) [inline dspace-api/src/main/java/org/dspace/core/Context.java]: That's a good and valid point. I think the easiest way is to clear the cache when switching to READ_ONLY. I'll try to add that to this PR.

File: dspace-api/src/main/java/org/dspace/core/HibernateDBConnection.java
Comment (tomdesair) [inline dspace-api/src/main/java/org/dspace/core/HibernateDBConnection.java]: This is null safe. `Hibernate.isInitialized()` will return false in case it receives a `null` object.

File: dspace-api/src/main/java/org/dspace/harvest/OAIHarvester.java
Comment (tomdesair) [inline dspace-api/src/main/java/org/dspace/harvest/OAIHarvester.java]: The difficulty with committing the transaction is that any reference to an entity from before the commit has to be reloaded from the database if we still intend to use them. This also applies to references in the calling method.

File: dspace/config/hibernate-ehcache-config.xml
Comment (tomdesair) [inline dspace/config/hibernate-ehcache-config.xml]: Valid point, I'll try to fine-tune those numbers.

**Comment by tomdesair** [comment]
Hi @pnbecker , commits https://github.com/DSpace/DSpace/pull/1694/commits/3b262d0a5eec4ab4542580409ef82c67489af185 and https://github.com/DSpace/DSpace/pull/1694/commits/aea353998f2880a72efb813ee95d8e4b7d083488 address your comments.

**Comment by pnbecker** [comment]
This is ready to be merged. As it needs to be cherry-picked to master too and my time is quite limited today, I delay merging by another day. If any other committer has time, I would appreciate if you'd merge and cherry-pick it.

**Comment by hardyoyo** [comment]
This needs to be forward-ported to the master branch, however, it's not quite a simple cherry-pick and push,  I've tried, there are conflicts. This would be a nice enhancement for the master branch's API as well... so, I think we probably need to create a Jira issue for that work.

**Comment by hardyoyo** [comment]
@tomdesair, @tdonohue suggested that I ping you on this. We need to forward-port this change to the master branch, however, simply cherry-picking the commit leads to many merge conflicts. Is there any chance you can make a PR for the master branch?

**Pull request merged**
